### PR TITLE
Add Deng shallow convection scheme (shcu_physics=5)

### PR DESCRIPTION
**Pull request by dudhia**
TYPE: new feature

KEYWORDS: Deng shallow convection added, shcu_physics=5

SOURCE: A.-J. Deng (PSU) via Pedro Jimenez (RAL)

DESCRIPTION OF CHANGES: 
[Text copied from Jimenez et al. (2016) WRF-Solar BAMS paper]. New shallow convection option previously implemented in the MM5 mesoscale model (Deng et al. 2003), which provides a cloud fraction for radiation. The Deng et al. (2003, 2014) shallow convection scheme is a mass-flux-based scheme. It includes a cloud entraining/detraining model to represent updrafts and it is triggered by factors including planetary boundary layer depth and turbulent kinetic energy (TKE). It uses a hybrid closure combining TKE and convective available potential energy, depending on the updraft depth. In addition to the updraft formulation, the scheme also contains two predictive equations for cloud fraction and cloud liquid/ice water content for neutrally buoyant clouds (inactive clouds detrained from the active updraft core).
Deng, A., N. L. Seaman, and J. S. Kain, 2003: A shallow-convection parameterization for mesoscale models. Part I: Submodel description and preliminary applications. J. Atmos. Sci., 60, 34–56.
doi:10.1175/1520-0469(2003)060<0034:ASCPFM >2.0.CO;2.
Deng, A., B. J. Gaudet, J. Dudhia, and K. Alapaty, 2014: Implementation and evaluation of a new shallow convection scheme in WRF. 26th Conf. on Weather Analysis and Forecasting/22nd Conf. on Numerical Weather Prediction, Atlanta, GA, Amer. Meteor. Soc., 12.5. [Available online at https://ams.confex.com/ams/94Annual/webprogram/Paper236925.html.]

LIST OF MODIFIED FILES: 
M       Registry/Registry.EM_COMMON
M       Registry/Registry.NMM
M       Registry/registry.dimspec
M       dyn_em/module_first_rk_step_part1.F
M       dyn_em/start_em.F
M       main/depend.common
M       phys/Makefile
M       phys/module_physics_addtendc.F
M       phys/module_physics_init.F
M       phys/module_radiation_driver.F
M       phys/module_shallowcu_driver.F
A       phys/module_shcu_deng.F
M       run/README.namelist

TESTS CONDUCTED: 
 - [x] Passed small GNU-only regression test.
 - [x] Passed regression test with this option activated.

RELEASE NOTE: 
New Deng et al. (2003) shallow convection option (shcu_physics=5) from WRF-Solar suite.

### CODE REVIEW

**Comment by dudhia** [comment]
Note that module_ra_farms.F shows up in the diffs, but is not supposed to be added at this time. It was removed from the dengonly branch and does not show up as A for added in diffs.

File: Registry/Registry.EM_COMMON [hunk 1]
```diff
@@ -1307,6 +1307,32 @@ state    real  RQGSHTEN        ikj      misc        1         -      r        "R
 state    real  RQCNSHTEN       ikj      misc        1         -      r        "RQCNSHTEN"             "Q_CN TENDENCY DUE TO SHALLOW CUMULUS SCHEME"     "# kg-1 s-1"
 state    real  RQINSHTEN       ikj      misc        1         -      r        "RQINSHTEN"             "Q_IN TENDENCY DUE TO SHALLOW CUMULUS SCHEME"     "# kg-1 s-1"
 
+# Deng Shallow Convection Parameterization
+state    real  RDCASHTEN       ikj      misc        1         -      r       "RDCASHTEN"             "DCA TENDENCY DUE TO DENG SHALLOW CUMULUS SCHEME"    "Pa s-1"
+state    real  RQCDCSHTEN      ikj      misc        1         -      r        "RQCDCSHTEN"            "DCQC TENDENCY DUE TO DENG SHALLOW CUMULUS SCHEME"   "kg kg-1 s-1"
+state    real  CLDAREAA        ikj      dyn_em      1         -      r        "CLDAREAA"              "CLOUD FRACTION FROM DENG SCHEME" ""
```
Comment (weiwangncar) [inline Registry/Registry.EM_COMMON]: Can these be lined up?

File: Registry/Registry.EM_COMMON [hunk 2]
```diff
@@ -2445,6 +2471,9 @@ rconfig   logical scm_force_wind_largescale namelist,scm  1     .false.     rh
 rconfig   integer scm_force_skintemp      namelist,scm  1       0           rh   "scm_force_skin"       "SCM surface forcing by skin temperature 0=no 1=yes"      ""
 rconfig   integer scm_force_flux          namelist,scm  1       0           rh   "scm_force_flux"       "SCM surface forcing by surface fluxes 0=no 1=yes"      ""
 
+# PSU-DENG SHALLOW CU VARIABLES
+state     real  WSUBSID          k        misc          1       -           -    "WSUBSID"              "LARGE-SCALE SUBSIDENCE IN DENG SHCU SCHEME"    "m s-1"
```
Comment (weiwangncar) [inline Registry/Registry.EM_COMMON]: Should we place all the variables related to this scheme together and remove many "PSU-DENG ShCu"?

File: dyn_em/module_first_rk_step_part1.F [hunk 3]
```diff
@@ -402,6 +402,14 @@ SUBROUTINE first_rk_step_part1 (   grid , config_flags              &
...
      &         ,diffuse_frac=grid%diffuse_frac &           
      &         ,IS_CAMMGMP_USED = grid%is_CAMMGMP_used    &
+!
+! PSU-DENG
+!
+     &         ,cw_rad=grid%cw_rad                     &
+     &         ,shcu_physics=config_flags%shcu_physics                     &
+!
+! END PSU-DENG
+!
```
Comment (weiwangncar) [inline dyn_em/module_first_rk_step_part1.F]: remove PSU-DEND etc..

File: dyn_em/module_first_rk_step_part1.F [hunk 4]
```diff
@@ -1332,6 +1340,32 @@ SUBROUTINE first_rk_step_part1 (   grid , config_flags              &
      &             ,HFX=grid%hfx, QFX=grid%qfx                            &
      &             ,MP_PHYSICS=config_flags%mp_physics                    &
      &             ,pgcon=config_flags%sas_pgcon                          &
+!
+! PSU-DENG
+!
```
Comment (weiwangncar) [inline dyn_em/module_first_rk_step_part1.F]: remove PSU-DENG

File: dyn_em/start_em.F [hunk 5]
```diff
@@ -1166,6 +1166,16 @@ SUBROUTINE start_domain_em ( grid, allowed_to_read &
                       ,grid%massflux_EDKF, grid%entr_EDKF, grid%detr_EDKF                &
                       ,grid%thl_up,grid%thv_up,grid%rt_up                                &
                       ,grid%rv_up,grid%rc_up,grid%u_up,grid%v_up,grid%frac_up            &
+!
+! PSU-DENG
```
Comment (weiwangncar) [inline dyn_em/start_em.F]: remove PSU-DENG

File: phys/module_physics_addtendc.F [hunk 6]
```diff
@@ -1679,6 +1679,50 @@ SUBROUTINE phy_shcu_ten(config_flags,rk_step,n_moist,n_scalar,   &
...
+
+        if (P_QR .ge. PARAM_FIRST_SCALAR)                        &
+        CALL add_a2a(moist_tendf(ims,kms,jms,P_QR),RQRSHTEN,     &
+                config_flags,                                    &
+                ids,ide, jds, jde, kds, kde,                     &
+                ims, ime, jms, jme, kms, kme,                    &
+                its, ite, jts, jte, kts, kte                     )
+!
+! END PSU DENG
+!
```
Comment (weiwangncar) [inline phys/module_physics_addtendc.F]: remove PSU-DEND and empty lines.

File: phys/module_physics_init.F [hunk 7]
```diff
@@ -212,6 +212,18 @@ SUBROUTINE phy_init ( id, config_flags, DT, restart, zfull, zhalf,     &
...
+!
+                         ,RDCASHTEN, RQCDCSHTEN                 &
+                         ,cldareaa, cldareab, cldliqa, cldliqb  &
+                         ,ca_rad, cw_rad                        &
+                         ,pblmax, wub, ltopb, clddpthb, cldtopb &
+                         ,capesave, ainckfsa, radsave           &
+                         ,rainsh, rainshvb, kdcldtop, kdcldbas  & 
+                         ,xtime1, PBLHAVG, TKEAVG               &
+!
+! END PSU-DENG
```
Comment (weiwangncar) [inline phys/module_physics_init.F]: more to remove for PSU-DENG

File: phys/module_physics_init.F [hunk 8]
```diff
@@ -722,6 +734,30 @@ SUBROUTINE phy_init ( id, config_flags, DT, restart, zfull, zhalf,     &
    REAL,OPTIONAL, DIMENSION( ims:ime , jms:jme ) , &
           INTENT(OUT) ::    wm_temf
 
+!
+!  Deng Shallow Cu variables
+!
```
Comment (weiwangncar) [inline phys/module_physics_init.F]: remove "Deng Shallow Cu variables"

File: phys/module_physics_init.F [hunk 9]
```diff
@@ -3776,6 +3869,27 @@ SUBROUTINE shcu_init(STEPCU,CUDT,DT,RUSHTEN,RVSHTEN,RTHSHTEN,   &
            ids, ide, jds, jde, kds, kde,                             &
            ims, ime, jms, jme, kms, kme,                             &
            its, ite, jts, jte, kts, kte                              )
+
+!
+!  PSU-DENG
+!
```
Comment (weiwangncar) [inline phys/module_physics_init.F]: remove "PSU-DENG"

File: phys/module_radiation_driver.F [hunk 10]
```diff
@@ -148,6 +148,13 @@ SUBROUTINE radiation_driver (                                  &
               ,aer_asy_opt, aer_asy_val                                   &
               ,aod5502d, angexp2d, aerssa2d, aerasy2d                     &
               ,aod5503d                                                   &
+!
+! PSU-DENG
+!
```
Comment (weiwangncar) [inline phys/module_radiation_driver.F]: Here too.

**Comment by dudhia** [comment]
Thanks. I will post a set of changes with these clean-ups soon.

**Comment by dudhia** [comment]
That is good to know. Thanks.

File: Registry/Registry.EM_COMMON (hunk 1 above)
Comment (davegill) [reply Registry/Registry.EM_COMMON]: Why are some of the arrays misc and others dyn_em? They are not in the NMM Registry. We should just choose one here.

File: Registry/Registry.EM_COMMON [hunk 11]
```diff
@@ -2789,6 +2816,7 @@ package   g3shcuscheme    shcu_physics==1            -             -
...
 package   grimsshcuscheme shcu_physics==3            -             -
 package   nscvshcuscheme  shcu_physics==4            -             -
+package   dengshcuscheme  shcu_physics==5            -             state:RDCASHTEN,RQCDCSHTEN,CLDAREAA,CLDAREAB,CA_RAD,CW_RAD,CLDLIQA,CLDLIQB,CLDDPTHB,CLDTOPB,PBLMAX,WUB,RAINSHVB,CAPESAVE,RADSAVE,AINCKFSA,LTOPB,KDCLDTOP,KDCLDBAS,XTIME1,PBLHAVG,TKEAVG,W0AVG
```
Comment (davegill) [inline Registry/Registry.EM_COMMON]: @dudhia @weiwangncar 
Folks,
Is the last variable in the shcu_physics==5 package supposed to be W0AVG or WSUBID

Comment (davegill) [reply Registry/Registry.EM_COMMON]: I see that wsubid is just a 1d array - I'm fine now

File: phys/module_shallowcu_driver.F [hunk 12]
```diff
@@ -352,6 +397,59 @@ SUBROUTINE shallowcu_driver(                                       &
          OPTIONAL, INTENT(IN   ) ::                      hfx,qfx
    INTEGER, INTENT(IN), OPTIONAL ::                   mp_physics
    REAL,    INTENT(IN), OPTIONAL ::                        pgcon
+
+
+! PSU-DENG Shallow cu variables
+!
+   INTEGER,   INTENT(IN   ) ::           bl_pbl_physics
+   REAL, DIMENSION( ims:ime, kms:kme, jms:jme ),                 &
+                   !jkim: INTENT(IN   ) ::           w, qke
```
Comment (davegill) [inline phys/module_shallowcu_driver.F]: remove these old pieces of code that are added in as commented-out

File: phys/module_shallowcu_driver.F [hunk 13]
```diff
@@ -352,6 +397,59 @@ SUBROUTINE shallowcu_driver(                                       &
...
+                   INTENT(IN   ) ::         xlong
+
+   INTEGER, DIMENSION( ims:ime, jms:jme ),                 &
+                   INTENT(INOUT) ::                              &
+                                                    ltopb, kdcldtop, kdcldbas
+
+   REAL, DIMENSION( ims:ime, 1:100, jms:jme ),                 &
+                   INTENT(INOUT) ::                              &
+                                 ainckfsa
+ !mchen  REAL, DIMENSION( ims:ime, kms:kme, jms:jme ),                 &
```
Comment (davegill) [inline phys/module_shallowcu_driver.F]: remove these commented out array declarations

File: phys/module_shallowcu_driver.F [hunk 14]
```diff
@@ -508,6 +606,84 @@ SUBROUTINE shallowcu_driver(                                       &
...
       ENDIF
 
+!
+! PSU-DENG
+!
+  CASE (DENGSHCUSCHEME)
+      CALL wrf_debug(100,'in PSU-DENG Shallow Cu CPS')
+
+
+    !  IF( bl_pbl_physics == 11 ) THEN
```
Comment (davegill) [inline phys/module_shallowcu_driver.F]: remove all this commented-out looping and assignment

File: phys/module_shallowcu_driver.F [hunk 15]
```diff
@@ -508,6 +606,84 @@ SUBROUTINE shallowcu_driver(                                       &
...
+            IDS=ids,IDE=ide,JDS=jds,JDE=jde,KDS=kds,KDE=kde              &
+           ,IMS=ims,IME=ime,JMS=jms,JME=jme,KMS=kms,KME=kme              &
+           ,ITS=its,ITE=ite,JTS=jts,JTE=jte,KTS=kts,KTE=kte              &
+           ,DT=dt ,KTAU=itimestep ,DX=dx, xtime=xtime, gmt=gmt           &
+           ,XLV=XLV, XLS=XLS                                             &
+           ,XLV0=XLV0 ,XLV1=XLV1 ,XLS0=XLS0 ,XLS1=XLS1                   &
+           ,CP=CP, R=R_d ,G=G                                            &
+           ,SVP1=SVP1 ,SVP2=SVP2 ,SVP3=SVP3 ,SVPT0=SVPT0                 &
+           ,ADAPT_STEP_FLAG=ADAPT_STEP_FLAG, DSIGMA=dnw                  &
+           ,XLONG=xlong, HT=ht, PBLH=pblh   &
```
Comment (davegill) [inline phys/module_shallowcu_driver.F]: line things up, here and the next 10-15 lines there are a few that would be nice

**Comment by dudhia** [comment]
I am acting on nearly all of Dave's clean-up suggestions. For now I am only leaving the ifdef as it
is and the comments that show which shallow cumulus driver args belong to Deng.

File: phys/module_radiation_driver.F [hunk 16]
```diff
@@ -1450,6 +1456,27 @@ SUBROUTINE radiation_driver (                                  &
            ENDDO
         end if
      endif
+
+#if (EM_CORE==1)
```
[the remark this answers is not in the document]
Comment (davegill) [reply phys/module_radiation_driver.F]: I'm OK with the ifdef

**Comment by davegill** [comment]
@dudhia 
Jimy,
I "resolved" most of the comments from @weiwangncar and me. There are a few outstanding. Please review, and either address them or comment as to why not.

@smileMchen 
Ming,
Other than a vanilla regression test, were there any other tests conducted? If so, would you edit the PR's commit message an include that information.

File: Registry/Registry.EM_COMMON (hunk 2 above)
Comment (dudhia) [reply Registry/Registry.EM_COMMON]: done

**Comment by weiwangncar** [comment]
@dudhia In your PR, you also mentioned a 2014 reference, but it isn't listed in the PR.

File: Registry/Registry.EM_COMMON [hunk 17]
```diff
@@ -1307,6 +1307,33 @@ state    real  RQGSHTEN        ikj      misc        1         -      r        "R
 state    real  RQCNSHTEN       ikj      misc        1         -      r        "RQCNSHTEN"             "Q_CN TENDENCY DUE TO SHALLOW CUMULUS SCHEME"     "# kg-1 s-1"
 state    real  RQINSHTEN       ikj      misc        1         -      r        "RQINSHTEN"             "Q_IN TENDENCY DUE TO SHALLOW CUMULUS SCHEME"     "# kg-1 s-1"
 
+# Deng Shallow Convection Parameterization
+state    real  RDCASHTEN       ikj      misc        1         -      r        "RDCASHTEN"             "DCA TENDENCY DUE TO DENG SHALLOW CUMULUS SCHEME"    "Pa s-1"
+state    real  RQCDCSHTEN      ikj      misc        1         -      r        "RQCDCSHTEN"            "DCQC TENDENCY DUE TO DENG SHALLOW CUMULUS SCHEME"   "kg kg-1 s-1"
```
Comment (weiwangncar) [inline Registry/Registry.EM_COMMON]: @dudhia What are DCA and DCQC tendencies?

File: phys/module_shallowcu_driver.F [hunk 18]
```diff
@@ -508,6 +571,68 @@ SUBROUTINE shallowcu_driver(                                       &
...
+            tke_scr(I,K,J) = 0.5 * qke(I,K,J)
+            kth_scr(I,K,J) = exch_h(I,K,J)
+            bbls_scr(I,K,J) = el_pbl(I,K,J)
+         ENDDO
+         ENDDO
+         ENDDO
+      ELSE
+         WRITE(message,*) 'PSU DENG ShCu currently does not support PBL option: ', &
+                            bl_pbl_physics
+         CALL wrf_error_fatal ( message )
```
Comment (weiwangncar) [inline phys/module_shallowcu_driver.F]: @dudhia Should this be checked in check_a_mundo? What about Boulac, QNSE, and MYNN 3.5, GBM schemes?

File: Registry/Registry.EM_COMMON [hunk 19]
```diff
@@ -1307,6 +1307,33 @@ state    real  RQGSHTEN        ikj      misc        1         -      r        "R
...
+state    real  CLDLIQA         ikj      dyn_em      1         -      r        "CLDLIQA"               "CLOUD WATER CONTENT FROM DENG SCHEME" "kg kg-1"
+state    real  CLDLIQB         ikj      dyn_em      1         -      r        "CLDLIQB"               "CLOUD WATER CONTENT FROM DENG SCHEME" "kg kg-1"
+state    real  CLDDPTHB         ij      misc        1         -      r        "CLDDPTHB"              "CLOUD DEPTH FROM TAU-1"                "m"
+state    real  CLDTOPB          ij      misc        1         -      r        "CLDTOPB"               "CLOUD TOP FROM TAU-1"                "m"
+state    real  PBLMAX           ij      misc        1         -      r        "PBLMAX"                "MAXIMUM VALUE OF PBL DEPTH SINCE THE START OF THE MODEL RUN"                 "m"
+state    real  WUB             ikj      misc        1         -      r        "WUB"                   "UPDRAFT VERTICAL VELOCITY"                 "m s-1"
+state    real  RAINSHVB         ij      misc        1         -      r        "RAINSHVB"              "TIME-STEP SHCU PRECIPITATION FROM TAU-1"       "m s-1"
```
Comment (weiwangncar) [inline Registry/Registry.EM_COMMON]: @dudhia Is there a reason to use a different time-step RAINSH or RAINSHV? If so, should the rain be considered in the time-step total that goes into LSM?

File: phys/module_shallowcu_driver.F (hunk 18 above)
Comment (davegill) [reply phys/module_shallowcu_driver.F]: @weiwangncar @dudhia 
Folks,
Let's make this an issue. There will be tweaking of the code, and check_a_mundo is among those modifications that can happen over the next several weeks.

File: Registry/Registry.EM_COMMON (hunk 19 above)
Comment (dudhia) [reply Registry/Registry.EM_COMMON]: Not clear rainshvb is used for anything.
rainshv is used for the LSM

File: dyn_em/start_em.F [hunk 20]
```diff
@@ -1166,6 +1166,13 @@ SUBROUTINE start_domain_em ( grid, allowed_to_read &
                       ,grid%massflux_EDKF, grid%entr_EDKF, grid%detr_EDKF                &
                       ,grid%thl_up,grid%thv_up,grid%rt_up                                &
                       ,grid%rv_up,grid%rc_up,grid%u_up,grid%v_up,grid%frac_up            &
+                      ,grid%RDCASHTEN, grid%RQCDCSHTEN                                   &
+                      ,grid%cldareaa, grid%cldareab, grid%cldliqa, grid%cldliqb          &
+                      ,grid%ca_rad, grid%cw_rad                                          &
+                      ,grid%pblmax, grid%wub, grid%ltopb, grid%clddpthb, grid%cldtopb    &
+                      ,grid%capesave, grid%ainckfsa, grid%radsave                        &
+                      ,grid%rainsh, grid%rainshvb, grid%kdcldtop, grid%kdcldbas          &
+                      ,grid%xtime1, grid%PBLHAVG, grid%TKEAVG                            &
```
[the remark this answers is not in the document]
Comment (dudhia) [reply dyn_em/start_em.F]: It seems to initialize most or all of these to 0 in the deng init part

File: phys/module_shallowcu_driver.F (hunk 18 above)
Comment (dudhia) [reply phys/module_shallowcu_driver.F]: OK, we can deal with it later

File: Registry/Registry.EM_COMMON (hunk 19 above)
Comment (dudhia) [reply Registry/Registry.EM_COMMON]: may want to remove it later

**Pull request merged**